File: golang_demo/http_consumer_demo.go
package main

import (
	"bytes"
	"cubetran_udf_golang/metadata"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/linkedin/goavro/v2"
)

type InfoResp struct {
	AckedBatchID uint64 `json:"acked_batch_id"`
	SentBatchID  uint64 `json:"sent_batch_id"`
}

type FetchResp struct {
	Data    [][]byte `json:"data"`
	BatchID uint64   `json:"batch_id"`
}

type AckResp struct {
	AckedBatchID uint64 `json:"acked_batch_id"`
}

type AckReq struct {
	AckBatchID uint64 `json:"ack_batch_id"`
}

const URL string = "http://127.0.0.1:10231"
const BATCH_SIZE uint = 2

// extract data from HTTP server and decode
func runHttpDemo() {
	codec, _ := goavro.NewCodec(metadata.AvroSchema)

	// api: info
	url := fmt.Sprintf("%s/info", URL)
	infoResp := testInfo(url)

	ackBatchId := infoResp.SentBatchID
	for {
		// api: fetch_new
		url = fmt.Sprintf("%s/fetch_new?batch_size=%d&ack_batch_id=%d", URL, BATCH_SIZE, ackBatchId)
		fetchResp := testFetch(url, codec)
		ackBatchId = fetchResp.BatchID

		if len(fetchResp.Data) == 0 {
			time.Sleep(5 * time.Second)
		}
	}
}

// extract data from HTTP server and decode
func runHttpDemo2() {
	codec, _ := goavro.NewCodec(metadata.AvroSchema)

	for {
		// api: info
		url := fmt.Sprintf("%s/info", URL)
		testInfo(url)

		// api: fetch_new
		url = fmt.Sprintf("%s/fetch_new?batch_size=%d", URL, BATCH_SIZE)
		fetchResp := testFetch(url, codec)

		// api: fetch_old
		url = fmt.Sprintf("%s/fetch_old?old_batch_id=%d", URL, fetchResp.BatchID)
		testFetch(url, codec)

		// api: ack
		url = fmt.Sprintf("%s/ack", URL)
		ackReq := AckReq{
			AckBatchID: fetchResp.BatchID,
		}
		testAck(url, ackReq)

		if len(fetchResp.Data) == 0 {
			time.Sleep(5 * time.Second)
		}
	}
}

func testInfo(url string) InfoResp {
	infoResp := sendGet[InfoResp](url)

	jsonData, _ := json.Marshal(infoResp)
	fmt.Printf("info response: %s\n", jsonData)

	return infoResp
}

func testFetch(url string, codec *goavro.Codec) FetchResp {
	fetchResp := sendGet[FetchResp](url)

	fmt.Printf("fetched batch_id: %d\n", fetchResp.BatchID)
	for _, item := range fetchResp.Data {
		apeDtsRecord := bytes_to_struct(codec, item)
		jsonData, _ := json.Marshal(apeDtsRecord)
		fmt.Printf("ape_dts record: %s\n", jsonData)
	}

	return fetchResp
}

func testAck(url string, ackReq AckReq) AckResp {
	ackResp := sendPost[AckReq, AckResp](url, ackReq)

	jsonData, _ := json.Marshal(ackResp)
	fmt.Printf("ack response: %s\n", jsonData)

	return ackResp
}

func sendGet[T any](url string) T {
	fmt.Printf("HTTP GET url: %s\n", url)

	// send request
	resp, err := http.Get(url)
	if err != nil {
		log.Fatalf("Failed to send GET request: %v", err)
	}
	defer resp.Body.Close()

	return parseResponse[T](resp)
}

func sendPost[T1 any, T2 any](url string, payload T1) T2 {
	fmt.Printf("HTTP POST url: %s\n", url)

	jsonData, err := json.Marshal(payload)
	if err != nil {
		log.Fatalf("Error marshaling JSON: %v", err)
	}

	resp, err := http.Post(url, "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		log.Fatalf("Error sending POST request: %v", err)
	}
	defer resp.Body.Close()

	return parseResponse[T2](resp)
}

func parseResponse[T any](resp *http.Response) T {
	// read response
	if resp.StatusCode != http.StatusOK {
		log.Fatalf("Received non-OK response: %s", resp.Status)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Fatalf("Failed to read response body: %v", err)
	}

	// parse response
	var result T
	err = json.Unmarshal(body, &result)
	if err != nil {
		log.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	return result
}

File: rust_demo/src/http_consumer_demo.rs
use anyhow::bail;
use reqwest::blocking::{Client, Response};
use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::thread;
use std::time::Duration;

use crate::metadata::avro_converter::AvroConverter;

const URL: &str = "http://127.0.0.1:10231";
const BATCH_SIZE: u64 = 2;

#[derive(Serialize, Deserialize, Debug)]
struct InfoResp {
    acked_batch_id: u64,
    sent_batch_id: u64,
}

#[derive(Serialize, Deserialize, Debug)]
struct FetchResp {
    data: Vec<Vec<u8>>, // Assuming data is a vector of byte vectors
    batch_id: u64,
}

#[derive(Serialize, Deserialize, Debug)]
struct AckReq {
    ack_batch_id: u64,
}

#[derive(Serialize, Deserialize, Debug)]
struct AckResp {
    acked_batch_id: u64,
}

pub fn run_http_demo() -> anyhow::Result<()> {
    let avro_converter = AvroConverter::new();
    let client = Client::new();

    let url = format!("{}/info", URL);
    let info_resp = test_info(&client, &url)?;

    let mut ack_batch_id = info_resp.sent_batch_id;
    loop {
        let url = format!(
            "{}/fetch_new?batch_size={}&ack_batch_id={}",
            URL, BATCH_SIZE, ack_batch_id
        );

        let fetch_resp = test_fetch(&client, &avro_converter, &url)?;
        ack_batch_id = fetch_resp.batch_id;

        if fetch_resp.data.is_empty() {
            println!("No data fetched, sleeping for 5 seconds...");
            thread::sleep(Duration::from_secs(5));
        }
    }
}

pub fn run_http_demo_2() -> anyhow::Result<()> {
    let avro_converter = AvroConverter::new();
    let client = Client::new();

    loop {
        // api: info
        let mut url = format!("{}/info", URL);
        test_info(&client, &url)?;

        // api: fetch_new
        url = format!("{}/fetch_new?batch_size={}", URL, BATCH_SIZE);
        let fetch_resp = test_fetch(&client, &avro_converter, &url)?;

        // api: fetch_old
        url = format!("{}/fetch_old?old_batch_id={}", URL, fetch_resp.batch_id);
        test_fetch(&client, &avro_converter, &url)?;

        // api: ack
        url = format!("{}/ack", URL);
        let ack_req: AckReq = AckReq {
            ack_batch_id: fetch_resp.batch_id,
        };
        test_ack(&client, &url, &ack_req)?;

        if fetch_resp.data.is_empty() {
            thread::sleep(Duration::from_secs(5));
        }
    }
}

fn test_info(client: &Client, url: &str) -> anyhow::Result<InfoResp> {
    let info_resp: InfoResp = send_get(client, url)?;
    println!("info response: {:?}", info_resp);
    Ok(info_resp)
}

fn test_fetch(
    client: &Client,
    avro_converter: &AvroConverter,
    url: &str,
) -> anyhow::Result<FetchResp> {
    let mut fetch_resp: FetchResp = send_get(client, url)?;

    println!("fetched batch_id: {}", fetch_resp.batch_id);
    for item in fetch_resp.data.iter_mut() {
        let ape_dts_record = avro_converter.avro_to_ape_dts(item)?;
        println!("ape_dts record: {}", json!(ape_dts_record))
    }

    Ok(fetch_resp)
}

fn test_ack(client: &Client, url: &str, ack_req: &AckReq) -> anyhow::Result<AckResp> {
    let ack_resp: AckResp = send_post(client, url, ack_req)?;
    println!("ack response: {}", json!(ack_resp));
    Ok(ack_resp)
}

fn send_get<T: for<'de> Deserialize<'de>>(client: &Client, url: &str) -> anyhow::Result<T> {
    println!("HTTP GET url: {}", url);
    let response: Response = client.get(url).send()?;
    parse_response(response)
}

fn send_post<T1: Serialize, T2: for<'de> Deserialize<'de>>(
    client: &Client,
    url: &str,
    payload: &T1,
) -> anyhow::Result<T2> {
    println!("HTTP POST url: {}", url);
    // Send the POST request
    let resp = client
        .request(Method::POST, url)
        .header("Content-Type", "application/json")
        .body(json!(payload).to_string())
        .send()?;
    // Parse the response
    parse_response::<T2>(resp)
}

fn parse_response<T: for<'de> Deserialize<'de>>(resp: Response) -> anyhow::Result<T> {
    if !resp.status().is_success() {
        bail!(format!(
            "bad http request, status: {}, resp: {:?}",
            resp.status(),
            resp.text()?
        ))
    }
    let result: T = serde_json::from_str(&resp.text()?)?;
    Ok(result)
}
